packages/data-transport-layer/src/da/blob/l1-beacon-client.ts

```diff
@@ -0,0 +1,71 @@
+import axios from 'axios'
+import { Blob } from './blob'
+
+export class L1BeaconClient {
+  private readonly baseUrl: string
+
+  constructor(baseUrl: string) {
+    this.baseUrl = baseUrl
+  }
+
+  // checks the beacon chain version, usually just use this as a ping method
+  async checkVersion(): Promise<void> {
+    const response = await axios.get(`${this.baseUrl}/eth/v1/node/version`)
+  }
+
+  // retrieve blobs from the beacon chain
+  async getBlobs(timestamp: number, indices: number[]): Promise<any[]> {
+    // calculate the beacon chain slot from the given timestamp
+    const slot = (await this.getTimeToSlotFn())(timestamp)
+    const sidecars = await this.getBlobSidecars(slot, indices)
+    const blobs = sidecars.map((sidecar: any) => {
+      const blob = new Blob(sidecar.blob)
+      return {
+        data: blob.toData(),
+        kzgCommitment: sidecar.kzg_commitment,
+        kzgProof: sidecar.kzg_proof,
+      }
+    })
+    return blobs
+  }
+
+  // retrieve blob sidecars from the beacon chain
+  async getBlobSidecars(slot: number, indices: number[]): Promise<any[]> {
+    const response = await axios.get(
+      `${this.baseUrl}/eth/v1/beacon/blob_sidecars/${slot}`,
+      {
+        params: { indices: indices.join(',') },
+      }
+    )
+    return response.data.data
+  }
+
+  // calculate the slot number from a given timestamp
+  async getTimeToSlotFn(): Promise<(timestamp: number) => number> {
```

Comment:
I'm not sure it's correct.

can you prove it on sepolia and mainnet?

Reply:
> I'm not sure it's correct.
> 
> can you prove it on sepolia and mainnet?

Sure, example on Sepolia. You can also verify with curl by yourself

```bash
curl https://eth-beacon-chain-sepolia.drpc.org/rest/eth/v1/beacon/genesis | jq .

{
  "data": {
    "genesis_time": "1655733600",
    "genesis_validators_root": "0xd8ea171f3c94aea21ebc42a1ed61052acf3f9209c00e4efbaaddac09ed9b8078",
    "genesis_fork_version": "0x90000069"
  }
}
```

```bash
curl https://eth-beacon-chain-sepolia.drpc.org/rest/eth/v1/config/spec | jq .

{
  "data": {
    ...
    "SECONDS_PER_SLOT": "12",
    ...
  }
}
```

From the above 2 responses, the genesis time is `1655733600` and slot is `12`.

The time of slot 6411882 is 1732676184, you can check it out [here](https://sepolia.beaconcha.in/slot/6411882).

`(1732676184 - 1655733600) / 12 = 6411882.91666667` , rounded with floor, which is 6411882.

Reply:
Also want to share the original go code from OP:

```
// GetTimeToSlotFn returns a function that converts a timestamp to a slot number.
func (cl *L1BeaconClient) GetTimeToSlotFn(ctx context.Context) (TimeToSlotFn, error) {
	cl.initLock.Lock()
	defer cl.initLock.Unlock()
	if cl.timeToSlotFn != nil {
		return cl.timeToSlotFn, nil
	}

	genesis, err := cl.cl.BeaconGenesis(ctx)
	if err != nil {
		return nil, err
	}

	config, err := cl.cl.ConfigSpec(ctx)
	if err != nil {
		return nil, err
	}

	genesisTime := uint64(genesis.Data.GenesisTime)
	secondsPerSlot := uint64(config.Data.SecondsPerSlot)
	if secondsPerSlot == 0 {
		return nil, fmt.Errorf("got bad value for seconds per slot: %v", config.Data.SecondsPerSlot)
	}
	cl.timeToSlotFn = func(timestamp uint64) (uint64, error) {
		if timestamp < genesisTime {
			return 0, fmt.Errorf("provided timestamp (%v) precedes genesis time (%v)", timestamp, genesisTime)
		}
		return (timestamp - genesisTime) / secondsPerSlot, nil
	}
	return cl.timeToSlotFn, nil
}
```

Reply:
Okay, I found the spec 

https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#compute_timestamp_at_slot

by the way, you can use the cache.

Reply:
the result should not be a float number.

`(1732676184 - 1655733600) / 12 == 6411882`

Reply:
> the result should not be a float number.
> 
> `(1732676184 - 1655733600) / 12 == 6411882`

Yeah I know, in go we use integer for this, so this is automatically rounded. But in NodeJS, number is a float by default, so we need to round it with `Math.floor` .

Reply:
> Okay, I found the spec
> 
> https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#compute_timestamp_at_slot
> 
> by the way, you can use the cache.

Got it, I'll add the cache.

Reply:
> Okay, I found the spec
> 
> https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#compute_timestamp_at_slot
> 
> by the way, you can use the cache.

Cache added in this [commit](https://github.com/MetisProtocol/mvm/pull/104/commits/16ca5630be3ac30ac123c12117c29341e5f43f54)